src/dtx/dtx_common.c

```diff
@@ -0,0 +1,528 @@
+	if (commit_dtes == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	D_ALLOC_ARRAY(abort_dtes, count);
+	if (abort_dtes == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	for (i = 0; i < count; i++) {
+		int	j;
+		bool	skip = false;
+
+		for (j = 0; j < i; j++) {
+			if (daos_dti_eq(&dces[i].dce_xid, &dces[j].dce_xid)) {
+				skip = true;
+				break;
+			}
+		}
+
+		if (!skip) {
+			rc = vos_dtx_lookup_cos(coh, oid, &dces[i].dce_xid,
+						dces[i].dce_dkey, true);
+			if (rc == -DER_NONEXIST)
+				rc = vos_dtx_lookup_cos(coh, oid,
+						&dces[i].dce_xid,
+						dces[i].dce_dkey, false);
+			if (rc == 0) {
+				daos_dti_copy(&commit_dtes[commit_cnt].dte_xid,
+					      &dces[i].dce_xid);
+				commit_dtes[commit_cnt++].dte_oid = *oid;
+				continue;
+			}
+
+			if (rc == -DER_NONEXIST) {
+				daos_dti_copy(&abort_dtes[abort_cnt].dte_xid,
+					      &dces[i].dce_xid);
+				abort_dtes[abort_cnt++].dte_oid = *oid;
+				continue;
+			}
+
+			goto out;
+		}
+	}
+
+	if (commit_cnt > 0)
+		rc = dtx_commit(po_uuid, co_uuid, commit_dtes, commit_cnt,
+				version);
+
+	if (abort_cnt > 0) {
+		rc = dtx_abort(po_uuid, co_uuid, abort_dtes, abort_cnt,
+				version);
+		if (rc == -DER_NONEXIST)
+			rc = 0;
+	}
+
+out:
+	if (commit_dtes != NULL)
+		D_FREE(commit_dtes);
+	if (abort_dtes != NULL)
+		D_FREE(abort_dtes);
+
+	return rc;
+}
+
+int
+dtx_batched_commit_register(struct ds_cont_hdl *hdl)
+{
+	struct ds_cont			*cont = hdl->sch_cont;
+	struct dtx_batched_commit_args	*dbca;
+	d_list_t			*head;
+
+	D_ASSERT(cont != NULL);
+
+	head = &dss_get_module_info()->dmi_dtx_batched_list;
+	d_list_for_each_entry(dbca, head, dbca_link) {
+		if (uuid_compare(dbca->dbca_cont_hdl->sch_cont->sc_uuid,
+				 cont->sc_uuid) == 0)
+			return 0;
+	}
+
+	D_ALLOC_PTR(dbca);
+	if (dbca == NULL)
+		return -DER_NOMEM;
+
+	cont->sc_closing = 0;
+	ds_cont_hdl_get(hdl);
+	dbca->dbca_cont_hdl = hdl;
+	d_list_add_tail(&dbca->dbca_link, head);
+
+	return 0;
+}
+
+void
+dtx_batched_commit_degister(struct ds_cont_hdl *hdl)
+{
+	struct ds_cont			*cont = hdl->sch_cont;
+	struct dtx_batched_commit_args	*dbca;
+	d_list_t			*head;
+
+	if (cont == NULL)
+		return;
+
+	if (cont->sc_closing) {
+		while (!cont->sc_dtx_flushed)
+			ABT_thread_yield();
+
+		return;
+	}
+
+	head = &dss_get_module_info()->dmi_dtx_batched_list;
+	d_list_for_each_entry(dbca, head, dbca_link) {
+		if (uuid_compare(dbca->dbca_cont_hdl->sch_cont->sc_uuid,
+				 cont->sc_uuid) == 0) {
+			D_ASSERT(hdl == dbca->dbca_cont_hdl);
+
+			cont->sc_dtx_flushed = 0;
+			cont->sc_closing = 1;
+
+			do {
+				ABT_thread_yield();
+			} while (!cont->sc_dtx_flushed);
+
+			return;
+		}
+	}
+}
+
+int
+dtx_handle_resend(daos_handle_t coh, daos_unit_oid_t *oid,
+		  struct dtx_id *dti, uint64_t dkey_hash,
+		  uint64_t start_time, bool punch)
+{
+	int	rc;
+
+	rc = vos_dtx_check_committable(coh, oid, dti, dkey_hash, punch);
+	switch (rc) {
+	case DTX_ST_PREPARED:
+		return 0;
+	case DTX_ST_INIT:
+		if (dti->dti_sec >= start_time)
+			return 0;
```

Comment:
seems un-trustable for the time/hlc comparison? maybe can just abort it as below

Reply:
Both the dti_sec and start_time are not HLC, they are wall clock. It is true that above comparing is not trustable, but we cannot abort it, that will cause the RPC abort itself and trigger resend again. Currently, since we do not support server re-integration, we can assume that the client RPC is always happened after server restart, right? so we can remove above check.